packages/labs/observers/src/mutation_controller.ts

```diff
@@ -0,0 +1,90 @@
+/**
+ * @license
+ * Copyright 2017 Google LLC
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+import {ReactiveControllerHost} from '@lit/reactive-element/reactive-controller.js';
+
+export type MutationValueCallback = (
+  ...args: Parameters<MutationCallback>
+) => unknown;
+
+export interface MutationControllerConfig {
+  config: MutationObserverInit;
+  target?: Element;
+  callback?: MutationValueCallback;
+  skipInitial?: boolean;
+}
+
+export class MutationController {
+  private _host: ReactiveControllerHost;
+  private _target?: Element | null;
+  private _config: MutationObserverInit;
+  private _observer: MutationObserver;
+  private _skipInitial = false;
+  private _unobservedUpdate = false;
+  value?: unknown;
+  callback: MutationValueCallback = () => true;
+  constructor(
+    host: ReactiveControllerHost,
+    {target, config, callback, skipInitial}: MutationControllerConfig
+  ) {
+    this._host = host;
+    this._target = target;
+    this._config = config;
+    if (skipInitial !== undefined) {
+      this._skipInitial = skipInitial;
+    }
+    if (callback !== undefined) {
+      this.callback = callback;
+    }
+    this._observer = new MutationObserver((records: MutationRecord[]) => {
+      this.handleChanges(records);
+      this._host.requestUpdate();
+    });
+    this._host.addController(this);
+  }
+
+  protected handleChanges(records: MutationRecord[]) {
+    this.value = this.callback(records, this._observer);
+  }
+
+  hostConnected() {
+    if (this._target) {
+      this.observe(this._target);
+    }
+  }
+
+  hostDisconnected() {
+    this.disconnect();
+  }
+
+  async hostUpdated() {
+    // Eagerly deliver any changes that happened during update.
+    // And handle initial state as a set of 0 changes. This helps setup initial
+    // state and promotes UI = f(state) since ideally the callback does not
+    // rely on changes.
+    const pendingRecords = this._observer.takeRecords();
+    if (
+      pendingRecords.length ||
+      (!this._skipInitial && this._unobservedUpdate)
+    ) {
+      this.handleChanges(pendingRecords);
+    }
+    this._unobservedUpdate = false;
+  }
+
+  /**
+   * Observe the target element.
+   * @param target Element to observe
+   */
+  observe(target: Element) {
+    this._observer.observe(target, this._config);
+    this._unobservedUpdate = true;
+    this._host.requestUpdate();
```

Comment:
you have requestUpdate here but not on intersection... intentional?

Reply:
Yes, IntersectionObserver reports the initial state of intersection whereas the other observers do not report their initial state so we trigger an update to do so.